### PR TITLE
refactor paymaster methods

### PR DESCRIPTION
**Pull request by himanshugarg06**

### CODE REVIEW

File: docs/Paymaster/paymastermethods.md [hunk 1]
```diff
@@ -5,39 +5,119 @@ sidebar_position: 3
...
+    mode: PaymasterMode.ERC20,
+    tokenList: ["0xdA5289FCAAF71d52A80A254dA614A192B693e975"],
+    preferredToken: "0xdA5289FCAAF71d52A80A254dA614A192B693e975",
+  },
+);
+```
+As per the code
+- We set mode to ERC20
+- In token list we can specify a list of addresses of the ERC20 tokens we want to have our users pay in.
+- We can also decide to choose a preferred token for the response to include.
```
Comment (NunoFilipeSantos) [inline docs/Paymaster/paymastermethods.md:38]:
```suggestion
As per the code
- We set the mode to ERC20
- In the token list we can specify a list of addresses of the ERC20 tokens we want to have our users pay in.
- We can also decide to choose a preferred token for the response to include.
```

File: docs/Paymaster/paymastermethods.md [hunk 2]
```diff
@@ -5,39 +5,119 @@ sidebar_position: 3
...
+
+### Parameters
+
+- userOp(`Partial<UserOperation>`, required): A UserOperation object representing the user's request that needs to be processed.
+- paymasterServiceData(`FeeQuotesOrDataDto`, required):  The paymaster service data containing token information and sponsorship details. you can send just the preferred token or array of token addresses.
```
Comment (NunoFilipeSantos) [inline docs/Paymaster/paymastermethods.md:43]:
```suggestion
- userOp(`Partial<UserOperation>`, required): A UserOperation object representing the user's request that needs to be processed.
- paymasterServiceData(`FeeQuotesOrDataDto`, required):  The paymaster service data containing token information and sponsorship details. You can send just the preferred token or an array of token addresses.
```

File: docs/Paymaster/paymastermethods.md [hunk 3]
```diff
@@ -5,39 +5,119 @@ sidebar_position: 3
...
+      preferredToken?: string; // If you want to pass only one token
+      webhookData?: {
+        [key: string]: any;
+      };
+      smartAccountInfo?: SmartAccountData; // use this to calulate the gas for smart account v1, otherwise by default it will return for latest smart account version
+    };
+    ```
+
+### Returns
+- response(`Promise<FeeQuotesOrDataResponse>`): It returns a promise that resolves to the following object containing a feeQuotes Array. A paymaster Fee quote will have additional information about the fee being paid in the specifed ERC20 token, how long this fee is valid, what type of premium is going to be paid, as well as general token information and amount in USD. Here is the full typing for a single fee quote:
```
Comment (NunoFilipeSantos) [inline docs/Paymaster/paymastermethods.md:61]:
```suggestion
- response(`Promise<FeeQuotesOrDataResponse>`): It returns a promise that resolves to the following object containing a feeQuotes Array. A paymaster Fee quote will have additional information about the fee being paid in the specified ERC20 token, how long this fee is valid, what type of premium is going to be paid, as well as general token information and the amount in USD. Here is the full typing for a single fee quote:
```

File: docs/Paymaster/paymastermethods.md [hunk 4]
```diff
@@ -5,39 +5,119 @@ sidebar_position: 3
...
+      premiumPercentage: number;
+      validUntil?: number;
+    };
+    ```
 
-### Mode: `SPONSORED`
+## getPaymasterAndData()
 
-Below is an example of the sponsorship data or `paymasterServiceData`
+The `getPaymasterAndData` method is used to get the information on how the transactionFees will be covered for the transactions. The returned `PaymasterAndDataResponse` includes a signed string (paymasterAndData) that signifies the paymaster's commitment to covering the transaction fee.
```
Comment (NunoFilipeSantos) [inline docs/Paymaster/paymastermethods.md:88]:
```suggestion
The `getPaymasterAndData` method is used to get the information on how the transaction fees will be covered for the transactions. The returned `PaymasterAndDataResponse` includes a signed string (paymasterAndData) that signifies the paymaster's commitment to covering the transaction fee.
```

File: docs/Paymaster/paymastermethods.md [hunk 5]
```diff
@@ -51,50 +131,28 @@ let paymasterServiceData: SponsorUserOperationDto = {
...
 
-```ts
-type PaymasterAndDataResponse = {
-  paymasterAndData: string;
-  preVerificationGas?: BigNumberish;
-  verificationGasLimit?: BigNumberish;
-  callGasLimit?: BigNumberish;
-};
-```
+ERC20 mode enables to pay the gas fees of your users in exchange for ERC-20 tokens. When switching mode to ERC20 there are additional steps that need to be considered.
```
Comment (NunoFilipeSantos) [inline docs/Paymaster/paymastermethods.md:145]:
```suggestion
ERC20 mode enables you to pay the gas fees of your users in exchange for ERC-20 tokens. When switching mode to ERC20 there are additional steps that need to be considered.
```

File: docs/Paymaster/paymastermethods.md [hunk 6]
```diff
@@ -5,39 +5,119 @@ sidebar_position: 3
...
-```
-
-After setting up your smart account as mentioned above you can start using the Paymaster in this way:
+import { IHybridPaymaster, PaymasterFeeQuote,
+ PaymasterMode, SponsorUserOperationDto} from "@biconomy/paymaster";
 
-```ts
 const biconomyPaymaster =
   smartAccount.paymaster as IHybridPaymaster<SponsorUserOperationDto>;
 ```
```
Comment (arcticfloyd1984) [inline docs/Paymaster/paymastermethods.md:17]: Would it make sense to split this section into two parts: `SPONSORED` and `ERC20`
Somone who knows which flow to use can just focus on that

File: docs/Paymaster/paymastermethods.md [hunk 7]
```diff
@@ -5,39 +5,119 @@ sidebar_position: 3
...
+  userOp,
+  {
+    mode: PaymasterMode.ERC20,
+    tokenList: ["0xdA5289FCAAF71d52A80A254dA614A192B693e975"],
+    preferredToken: "0xdA5289FCAAF71d52A80A254dA614A192B693e975",
+  },
+);
+```
+As per the code
+- We set mode to ERC20
```
Comment (arcticfloyd1984) [inline docs/Paymaster/paymastermethods.md:36]: I think it needs more clarity on what response is received based on mode that is passed
`ERC20` and `SPONOSRED` mode both return different responses

**Review by arcticfloyd1984 — CHANGES_REQUESTED** [review]

File: docs/Paymaster/paymastermethods.md (hunk 7 above)
Comment (himanshugarg06) [reply docs/Paymaster/paymastermethods.md]: yes, in the usage section of this method, will clearly specify usage for both modes

**Pull request closed without merging**